### PR TITLE
Merge `Rc::ptr_eq` on traits use into a SharedModel

### PR DESCRIPTION
**Pull request by hunger**
It is now safe to replace the manual implementation with the derived
implementation, so do that.

This is a continuation of #311, which I accidentally closed when I deleted the branch.

### CODE REVIEW

**Comment by ogoffart** [comment]
Is the title of this PR still accurate?

So the main thing this does, if i read correctly, is that it change `ModelHandle<T>` to be a `Option<SharedModel<T>>`  and that's a new class.
I don't remember the rational for changing the API there, but i don't understand why we need both ModelHandle and SharedModel.  SharedModel itself could wrap an `Option<Rc<dyn Model>>`  I believe we said we don't want to have the internal pub anyway. And so there should only be one of these class in the public API

**Comment by hunger** [comment]
The original reason to have this was to deduplicate the `Rc::ptr_eq` that we used to have and that clippy is not too happy abouti with trait object in the picture.

This puts all the code into one place that does `Rc::ptr_eq` on traits.

**Comment by ogoffart** [comment]
So what we said in the call is that we don't want to have `Option<SharedModel>`, but instead, we want `SharedModel` that internally contains an `Option<Rc<dyn Model>` but without being public.

I think we want conversion functions like so:
Currently, ModelHandle has a conversion `From<Option<Rc<...>>>` , but i think it should just be `From<Rc<...>>`.
We can also have a `TryInto<Rc<...>>`

Currently, `ModelHandle::new` Takes a Rc,  but i think it would be better if  it was `fn new(model: impl Model<T>)`
or maybe it could be `fn new(model: impl Into<Rc<dyn Model<Data = T>>)` 
The rational is that i want to be able to write code like this:
```rust
ModelHandle::new(VecModel::from_slice(&[...]))
```

Maybe we also want `impl <T, F: Model<Data= T>> From<F>  for SharedModel<T>`

**Comment by tronical** [comment]
> Maybe we also want impl <T, F: Model<Data= T>> From<F> for SharedModel<T>

I'd like that, too. On the one hand it's good to make the `Rc` allocation/wrapping visible to the user, on the other hand `SharedModel` implies the wrapping in a way and I've always found the `as Rc<dyn Model<Data = ...>>` part annoying to write. So 👍 for the above if that adds the convenience. I think it'll show pretty quickly in our own examples and tests whether it helps or not.

**Comment by ogoffart** [comment]
Regarding the `Shared` Terminology:
When we have, say:
```rust
let a = SharedString::new("hello");
let mut b = a.clone();
b.push_str("world");
// a didn't change
assert_eq!(a, "hello");
```

But that's not the case with the models:
```
let a = SharedModel::from(...);
let b = a.clone();
b.set_row_data(1, "foobar".into());
// a **is** changed
assert_eq!(a.row_data(1).unwrap(), "foobar".into());
```

Which is why the `Shared` terminology maybe isn't optimal.

Maybe `ModelRc<T>` would actually be a better name?

But if you think it's not a problem, I don't oppose `ModelShared<T>`

File: api/sixtyfps-rs/migration.md [hunk 1]
```diff
@@ -56,6 +56,22 @@ fn model_tracker(&self) -> &dyn ModelTracker {
 }
 ```
 
+#### `Option<SharedModel>` replaces `ModelHandle`
+
+`Option<SharedModel>` replaces `ModelHandle`.
+
+Old code:
+
+```rust
```
Comment (ogoffart) [inline api/sixtyfps-rs/migration.md:65]: I don't think you need to always put code sample for such trivial things like that.
Anyway, the ```rust must compile as it is going to be a doctest, so you need to add  `ignore` 
But i realize that this file is not included in the rust doc yet.

File: api/sixtyfps-rs/migration.md [hunk 2]
```diff
@@ -56,6 +56,22 @@ fn model_tracker(&self) -> &dyn ModelTracker {
 }
 ```
 
+#### `Option<SharedModel>` replaces `ModelHandle`
```
Comment (tronical) [inline api/sixtyfps-rs/migration.md:59]: This doesn't seem quite right, because in < 0.2.0 there's no `Option<SharedModel>`. I think perhaps you meant "ModelHandle was renamed to SharedModel"?

In that case it's a simple search and replace, if we don't change the API in a way that breaks existing code, just add new convenience (like the `From` from the model without Rc).

Comment (ogoffart) [reply api/sixtyfps-rs/migration.md:59]: > if we don't change the API in a way that breaks existing code

We will if we remove the Deref to the Option<Rc> and the pub access to .0, which we should.

File: sixtyfps_compiler/generator/rust.rs [hunk 3]
```diff
@@ -1398,11 +1398,12 @@ fn compile_expression(expr: &Expression, ctx: &EvaluationContext) -> TokenStream
             debug_assert!(matches!(array.ty(ctx), Type::Array(_)));
             let base_e = compile_expression(array, ctx);
             let index_e = compile_expression(index, ctx);
-            quote!(match &#base_e { x => {
+            quote!({
+                let x = &(#base_e);
```
Comment (ogoffart) [inline sixtyfps_compiler/generator/rust.rs:1402]: that's not the same.  We want to make sure that the temporaries in #base_e are not destroyed.

File: sixtyfps_compiler/generator/rust.rs [hunk 4]
```diff
@@ -1763,10 +1764,11 @@ fn compile_builtin_function_call(
         }
         BuiltinFunction::ImageSize => quote!( #(#a)*.size()),
         BuiltinFunction::ArrayLength => {
-            quote!(match &#(#a)* { x => {
+            quote!({
+                let x = &#(#a)*;
```
Comment (ogoffart) [inline sixtyfps_compiler/generator/rust.rs:1768]: ditto

File: sixtyfps_runtime/corelib/model.rs [hunk 5]
```diff
@@ -694,45 +688,40 @@ impl<C: RepeatedComponent> PinnedDrop for Repeater<C> {
...
-    fn model(self: Pin<&Self>) -> ModelHandle<C::Data> {
-        // Safety: Repeater does not implement drop and never let access model as mutable
+    fn model(self: Pin<&Self>) -> SharedModel<C::Data> {
+        // Safety: Repeater does not implement drop and never allows access to model as mutable
         let model = self.project_ref().model;
 
         if model.is_dirty() {
             *self.inner.borrow_mut() = RepeaterInner::default();
             self.is_dirty.set(true);
-            if let ModelHandle(Some(m)) = model.get() {
```
Comment (ogoffart) [inline sixtyfps_runtime/corelib/model.rs:704]: Too bad that we loose this optimisation that we won't allocate stuff if the model is empty.
I guess that's fine...
Otherwise you could use pub(crate)  to allow us to access the inner one.
Or we make TryInto Rc<> work

File: sixtyfps_runtime/corelib/model.rs [hunk 6]
```diff
@@ -769,11 +758,7 @@ impl<C: RepeatedComponent + 'static> Repeater<C> {
...
 
-        let model = if let ModelHandle(Some(model)) = self.model() {
-            model
-        } else {
-            return empty_model();
-        };
+        let model = self.model();
         let row_count = model.row_count();
         if row_count == 0 {
             return empty_model();
```
Comment (ogoffart) [inline sixtyfps_runtime/corelib/model.rs:764]: empty_model can now be inlined there, if i'm not mistaken.

File: sixtyfps_runtime/corelib/model.rs [hunk 7]
```diff
@@ -462,56 +462,50 @@ impl Model for bool {
     }
 }
 
-/// Properties of type array in the .60 language are represented as
-/// an [`Option`] of an [`Rc`] of something implemented the [`Model`] trait
+/// Properties of type Array in the .60 language are represented as
+/// a `SharedModel` that implements the `Model` trait.
+#[repr(transparent)]
 #[derive(derive_more::Deref, derive_more::DerefMut, derive_more::From, derive_more::Into)]
```
Comment (ogoffart) [inline sixtyfps_runtime/corelib/model.rs:468]: This needs to go away as well, but can be done in another commit

File: sixtyfps_runtime/corelib/model.rs [hunk 8]
```diff
@@ -462,56 +462,50 @@ impl Model for bool {
...
-pub struct ModelHandle<T>(pub Option<Rc<dyn Model<Data = T>>>);
+pub struct SharedModel<T>(Option<Rc<dyn Model<Data = T>>>);
 
-impl<T> core::fmt::Debug for ModelHandle<T> {
-    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
-        write!(f, "ModelHandle({:?})", self.0.as_ref().map(|_| "dyn Model"))
+impl<T> std::fmt::Debug for SharedModel<T> {
```
Comment (ogoffart) [inline sixtyfps_runtime/corelib/model.rs:471]:
```suggestion
impl<T> core::fmt::Debug for SharedModel<T> {
```

We want to compile without std on MCU, remember

**Review by ogoffart — CHANGES_REQUESTED** [review]

**Comment by ogoffart** [comment]
Maybe we can have a ` #[deprectated] type ModelHandle<T> = SharedModel<T>;`  to ease porting?

File: sixtyfps_compiler/generator/rust.rs (hunk 3 above)
Comment (hunger) [reply sixtyfps_compiler/generator/rust.rs:1402]: I do not get what you are going at, sorry.

**Comment by hunger** [comment]
Does the deprecation make sense in the transition from 0.1 to 0.2?

File: sixtyfps_compiler/generator/rust.rs (hunk 3 above)
Comment (ogoffart) [reply sixtyfps_compiler/generator/rust.rs:1402]:
```rust
let x = &String::new("42").trim();   // doesn't work (take address of temporary)
println!("{}", x); 
```
```rust
match (&String::new("42").trim()) {
  // works
  x => { println!("{}", x); }
```

So please undo that.

Why are you changing it?

Comment (ogoffart) [reply sixtyfps_compiler/generator/rust.rs:1402]: I guess something like this would fail to compile in rust:

```
property <string> world: (["hello", "world"])[1];
```

Comment (hunger) [reply sixtyfps_compiler/generator/rust.rs:1402]: Wow, tricky, but obvious now that you pointed it out!

File: sixtyfps_runtime/corelib/model.rs (hunk 5 above)
Comment (hunger) [reply sixtyfps_runtime/corelib/model.rs:704]: Will an empty model ever get dirty?

Comment (ogoffart) [reply sixtyfps_runtime/corelib/model.rs:704]: it can grow, yes.

Comment (ogoffart) [reply sixtyfps_runtime/corelib/model.rs:704]: I mean, what's an empty model.  a model with row_count = 0  but an actual model can grow.
But a model backed by None cannot change unless the property itself is changed. So no, a model with None cannot become dirty

Comment (hunger) [reply sixtyfps_runtime/corelib/model.rs:704]: So we do not loose the optimization: `None` will never get dirty, so it will never end up here in the first place.

File: api/sixtyfps-rs/migration.md [hunk 9]
```diff
@@ -62,3 +62,6 @@ fn model_tracker(&self) -> &dyn ModelTracker {
 }
 ```
 
+#### Minor changes to `Model`
+
+`ModelRc` replaces `ModelHandle`.
```
Comment (ogoffart) [inline api/sixtyfps-rs/migration.md:67]:
```suggestion
`ModelHandle` was renamed [`ModelRc`].
```

File: sixtyfps_runtime/corelib/model.rs [hunk 10]
```diff
@@ -462,56 +462,49 @@ impl Model for bool {
     }
 }
 
-/// Properties of type array in the .60 language are represented as
-/// an [`Option`] of an [`Rc`] of something implemented the [`Model`] trait
-#[derive(derive_more::Deref, derive_more::DerefMut, derive_more::From, derive_more::Into)]
-pub struct ModelHandle<T>(pub Option<Rc<dyn Model<Data = T>>>);
+/// Properties of type Array in the .60 language are represented as
+/// a `ModelRc` that implements the `Model` trait.
```
Comment (ogoffart) [inline sixtyfps_runtime/corelib/model.rs:466]:
```suggestion
/// a `ModelRc` that implements the [`Model`] trait.
```

File: sixtyfps_runtime/corelib/model.rs (hunk 5 above)
Comment (ogoffart) [reply sixtyfps_runtime/corelib/model.rs:704]: `model` in this case is a `Property<ModelRc>`  It's the property that gets dirty here, not the model.
But i guess we can live with that for now.

**Review by ogoffart — APPROVED** [review]

File: api/sixtyfps-rs/migration.md (hunk 9 above)
Comment (tronical) [reply api/sixtyfps-rs/migration.md:67]: renamed *to* :-)

**Review by tronical — APPROVED** [review]
I really like the name btw. This is a nice improvement of the API.

File: sixtyfps_runtime/corelib/model.rs [hunk 11]
```diff
@@ -462,56 +462,49 @@ impl Model for bool {
...
+#[derive(derive_more::Deref, derive_more::DerefMut)]
```
Comment (ogoffart) [inline sixtyfps_runtime/corelib/model.rs:467]: We still need to get rid of Deref and DerefMut, but we can do that in another PR
